File: src/services/share.ts
/**
 * Share links for content (documentaries, reels, inspire items, live slots).
 *
 * Deliberately mirrors payments.ts shareLinks() rather than introducing a
 * second share mechanism: same wa.me / sms: / mailto: shapes, same
 * encodeURIComponent handling. The difference is what is shared — a public
 * seo-site URL instead of a Razorpay payment link.
 *
 * Shared URLs point at seo-site (vallavan.in), NOT the SPA (app.vallavan.in).
 * That is the whole point: seo-site server-renders per-item OpenGraph tags, so
 * a pasted link shows a real title, description and thumbnail. The SPA is a
 * client-rendered bundle whose meta tags are identical on every route, so a
 * link to it would preview as the generic homepage.
 */

/** Public site base. Overridable at build time; defaults to the live domain. */
export const SITE_URL: string =
  ((import.meta.env.VITE_SITE_URL as string | undefined) || 'https://vallavan.in').replace(/\/$/, '');

/** Content kinds and the seo-site route each one maps to. */
export type ShareKind = 'documentary' | 'reel' | 'inspire' | 'live';

const ROUTES: Record<ShareKind, string> = {
  documentary: 'documentaries',
  reel: 'videos',
  inspire: 'inspire',
  live: 'live',
};

/** Canonical public URL for a piece of content. */
export function shareUrl(kind: ShareKind, id: string): string {
  return `${SITE_URL}/${ROUTES[kind]}/${id}`;
}

/** WhatsApp / SMS / email deep-links — same shape as payments.ts shareLinks(). */
export function contentShareLinks(url: string, title: string) {
  const msg = `${title} — watch on Vallavan: ${url}`;
  return {
    whatsapp: `https://wa.me/?text=${encodeURIComponent(msg)}`,
    sms: `sms:?body=${encodeURIComponent(msg)}`,
    email: `mailto:?subject=${encodeURIComponent(title)}&body=${encodeURIComponent(msg)}`,
  };
}

export type ShareOutcome = 'shared' | 'dismissed' | 'unsupported';

/**
 * Try the OS share sheet.
 *
 * Returns 'unsupported' when the browser has no Web Share API (every desktop
 * browser except Safari, and any non-secure context) so the caller can open
 * the WhatsApp/SMS/email fallback instead. A user who opens the sheet and
 * cancels raises AbortError — that is 'dismissed', not a failure, and must not
 * surface an error toast.
 */
export async function nativeShare(title: string, url: string): Promise<ShareOutcome> {
  if (typeof navigator === 'undefined' || typeof navigator.share !== 'function') return 'unsupported';
  try {
    await navigator.share({ title, text: title, url });
    return 'shared';
  } catch (err) {
    if ((err as Error)?.name === 'AbortError') return 'dismissed';
    // Any other failure (permission, unsupported payload) falls back to the menu.
    return 'unsupported';
  }
}

File: lib/utils/share.dart
/// Sharing content out of the app.
///
/// URLs point at seo-site (vallavan.in), NOT the app — the same rule the web
/// client follows in src/services/share.ts. seo-site server-renders per-item
/// OpenGraph tags, so a pasted link previews with a real title, description and
/// thumbnail; a link into the SPA would preview as the generic homepage because
/// its meta tags are identical on every route.
///
/// Deep links back into the app (App Links / Universal Links) are deliberately
/// out of scope for now, so a shared link opens in the browser.
library;

import 'package:share_plus/share_plus.dart';

import '../config/env.dart';

/// Content kinds and the seo-site route each maps to. Mirrors ShareKind in
/// src/services/share.ts — the two must agree or the clients emit different
/// URLs for the same row.
enum ShareKind { documentary, reel, inspire, live }

const Map<ShareKind, String> _routes = {
  ShareKind.documentary: 'documentaries',
  ShareKind.reel: 'videos',
  ShareKind.inspire: 'inspire',
  ShareKind.live: 'live',
};

/// Canonical public URL for a piece of content.
///
/// [Env.siteUrl] carries a trailing slash (it doubles as the Referer sent to
/// Bunny's CDN), so it is trimmed here rather than producing a double slash.
String shareUrlFor(ShareKind kind, String id) {
  final base = Env.siteUrl.replaceAll(RegExp(r'/+$'), '');
  return '$base/${_routes[kind]}/$id';
}

/// The message body. Same wording as the web client's contentShareLinks().
String shareMessageFor(String title, String url) => '$title — watch on Vallavan: $url';

/// Open the OS share sheet for a piece of content.
///
/// Uses `text` rather than `uri`: ShareParams documents that the two cannot be
/// combined, and text carries the title alongside the link on every platform,
/// where a bare uri would share a naked URL with no context. `subject` is used
/// by targets that have one, such as email.
Future<void> shareContent({
  required ShareKind kind,
  required String id,
  required String title,
}) async {
  final url = shareUrlFor(kind, id);
  await SharePlus.instance.share(
    ShareParams(text: shareMessageFor(title, url), subject: title),
  );
}
